### ISSUE TITLE
Openhab2 binding - Anyone already working on one?

### ISSUE THREAD

**Issue by javamatte** [homieiot/convention#10]
Hi, I'm an openhab user converting to openhab2 and an ESP8266 enthusiast. OH2 is based on the Eclipse SmartHome project, and their new binding architecture looks like it would mesh well with the Homie conventions. I've got a couple of Homie devices piped in using a direct MQTT binding, but a more robust, auto-discoverable binding could make adding these $10 devices a breeze!

Since OH2 now supports device discovery, the manual configuration process could be bypassed which could open up a new user base for homie-based devices. My MVP feature set would be auto-discovery (listen for new device/+/... topics) and categorization (switch, dimmer, read-only sensor) based on the Homie device's $nodes. There would have to be some possibly brittle mapping of $node:types to Openhab Item types to pull it off, but I don't think that's a show-stopper.

TL;DR:
My question is whether or not you know of anyone working on this yet... I'll gladly jump into another project if it exists! Otherwise, I'll link you the project/repo once I've got the binding started!

By the way, huge thanks for the great project! I've got an LED lighted flower arrangement that's been running Homie firmware for months. I've never had to restart it manually; it reattached to my MQTT broker even after I had to re-flash my RPi (it was down for days)... no problems at all!

**Comment by euphi** [comment]
I was thinking exactly the same. Homie is well fitted for OH2/Eclipse SmartHome auto-discovery, it's just necessary to give a more precise conventions regarding datatype/category.

I would love to support such a project - unfortunately I will be very busy at work for till at least April, so I'm only able to contribute with some proposals/reviews.

Do you have experience developing OH2 addons?

**Comment by marvinroger** [comment]
@javamatte it was in my plans to develop an openHAB2 binding at some point, but I really have 0 knowledge of openHAB.

So far I am not aware of anyone working on this.

Glad it has been working so stable for you! Thanks for wanting to contribute to the Homie ecosystem. I definitely could not bring the project this far if I was alone working on the project. 😄

**Comment by euphi** [comment]
A tutorial for binding development is available at https://www.eclipse.org/smarthome/documentation/development/bindings/how-to.html.

_Note: Eclipse Smarthome is the base of OpenHab2, so the bindings will be bindings for Eclipse Smarthome._

Eclipse Smarthome uses `Things `and `Items`.

When mapping this to Homie, each Homie-Property is an `Item` (with MQTT as the "`channel`" to link the `Item `to the `Thing`).

However, I'm unsure how to map an `Thing` to homie? Is each Homienode a `Thing`? Or each Homie-Device?

In my opinion, we should map each node to `Thing`. Thus, each Homie-Device may provide several `Things` for Eclipse Smarthome. All Homie-Nodes (=Things) of one Homie-Device may be grouped together.

What's your opinion about this?

**Comment by meverett-expedia** [comment]
_(FYI, I'm referencing this project and the Eclipse Smarthome stuff here: http://www.eclipse.org/smarthome/documentation/concepts/index.html (and sub pages) so we stay on the same page)_

ESH has an established (and probably growing/changing) list of Item types (from the Items page from the docs above), and they look like they could map fairly easily to Homie Nodes. In that scenario, I think the ESH/OH Thing should map to a homie Device. I'm still a bit fuzzy on the ESH/OH Channel. I'm lumping that in with the Node/Item for right now.

Discovery reveals new Things/Devices which provide 1 (or more!) Items/Nodes that are the actual actuators or sensors.

This is different from the mapping you described, but I'm definitely open to discussion on the topic! I think it's a key thing to get hammered out before we start slinging code.

Another Potential Issue:
I'm not sure how to require the MQTT broker; I'm not sure of OH has a dependency system yet. Clearly calling it out in the docs and having a sanity check prior to launching the AddOn should be sufficient for a rough cut.

**Comment by euphi** [comment]
### Mapping
Mapping each node to an item is not possible, if a node has more than one property.

According to the "pure" convention, this is not foressen; however there are many use case for Nodes with more than one property. See https://github.com/marvinroger/homie-esp8266/issues/124 for an example. Another example would be a Sensor that provides temperature and humidity. It makes sense to implement this as one node.
Also, if you make use of the "range" feature, each index will be an item.

I think, some clarification is necessary for the Homie convention.

### Channel
A channel seems to be just a presentation of the technical details of a Property. So in fact it would be the actual MQTT topics to send/receive.

### Broker
Regarding the MQTT broker: The MQTT *binding* requires to configure the broker address. I think this is fine, to have some configurable options for a "Homie" binding.  (MQTT broker adress, SSL support, User/Pass etc.).

Some configuration is neccesary for other bindings, too (e.g. the zwave binding requires the device name of the Z-Wave Stick). The Paper UI supports configuration.

**Comment by ordonot** [comment]
I really love the idea! :tada: :

The Homematic Binding is working well with openHAB 2 and the Paper UI. Would be nice to have the Homie Binding look similar. Without studying the documentation my understanding is

**Homie   |   ESH/OH**
device   |   Thing
device properties   |   Thing Configuration Parameters and Thing Properties
node   |   named group of channels*
node property   |   channel

In my point of view a ESH/OH item is something different. E.g. a Homie driven relay/MOSFET board offers some _switch and dimmer channels_ (not a light or a coffee machine). You can connect your LED strip (item) to one or even several channels. In my point of view this should be mapped in ESH/OH or Node-RED or somewhere like that but not in Homie itself.

\* Paper UI screenshot

![paperui](https://cloud.githubusercontent.com/assets/18089694/19668589/1b469c94-9a59-11e6-9672-e87de8e7c446.PNG)

**Comment by sja** [comment]
I read [this from Homie Readme.md](https://github.com/marvinroger/homie#background) and I think about the following mapping:

 Homie                        | ESH                                            | Comment
 -------------------- | ------------------------------ | --------------
 homie/mqtt               | Thing (implemented as Bridge Thing) |
 device                       | Thing                                          |
 device property        | Thing properties or Channel   | If just informational and read only, its a property.
 node                          | Channel                                    |
 node property          | Channel meta data                   | Unity types and so are defined in the xml as an attribute of a channel.
 settable property     | Channel meta data           | Also defined in xml. Those are rw instead of ro.


While researching, don't mix OH1 Bindings which run with compatibility layer wit newer ESH/OH2 bindings. Good references IMHO are Yahoo Weather (easy), Hue (normal) and Homematic (complex). 

I think the MQTT binding can be a reference/source, but don't depend on them. A binding should be independent, so should provide a mqtt lib. See Hue binding for that. It comes with a library for accessing the Hue bridge.

As a result, there should be a Homie Thing, which (if I understood that right) establishes a MQTT connection. With this, the devices can be discovered through the data available over that MQTT data. That also means that Homie itself cannot be auto discovered, but if there is a way to discover the MQTT server.

**Comment by sja** [comment]
My 2 cents:

Homie should be a BridgeThing with config props lile ip etc. 
A Homie node is a Thing with channels. Those channels should have types and can act on commands send to them. Usually those are forwarded through their BridgeThing. 

Sent from my mobile, so not very verbose. Looking forward for that binding. Let me know if you need details in modeling/mapping that. 

At first you should read the link, Ian mentioned. 

Von meinem Toaster gesendet

**Comment by RomeuCarvalhoPT** [comment]
Hello guys.
I'm very interested in this, is anyone working on a OH2 binding for this?
I was not aware of the existence of homie, but that was what I intended to build by myself.
Anyway, now that i know about homie, makes all the sense to build a OH2 binding for Homie, or even better, in a more generic way for MQTT.
I work in IT, but my knowledge of the framework(ESH/OH2/advanced java) is very limited and looks like it has a steep learning curve and my free time is not much.
Anyway I will continue my research on the subject and do what i can.
 
@sja I like your approach.
Probably to make things easier and more visible when you use it, i would remove the first layer (homie as a bridge thing). This way you can abstract from homie and the binding can be used by any other implementation that uses the same topic naming convention.
This way the binding auto discover service could be listening to all topics like "/devices/#" and for each new device publishing data, just create a new Thing.

I hope it makes sense.
I will keep you guy in formed in case i'm able to do, at least part of the binding.

**Comment by sja** [comment]
I think you cannot discover then. An MQTT connection is made from a Binding, not from the system itself.
You can either extend the MQTT Binding to understand Homie and then discover those, but even then you'll have a discovered Thing which handlers/logic is implemented in another, foreign binding. I think that this is not possible.
But I think it's not bad to have two MQTT-Connections from an openHAB to the broker. MQTT is made for many connections. So if you're running a MQTT broker just for Homie purposes, just install the openHAB Homie binding. If you're doing additional stuff with MQTT, you also need to install the 'normal' MQTT Binding to access them. (atm only available as oh 1 binding with manual config afaik)

**Comment by kwave** [comment]
I've implemented an openhab 2 binding for homie, which will be released to the public in a few days. It maps Homie Devices as Bridge, Homie Nodes as Things, and properties as channel. The binding supports all properties propagated by Homie devices. Works like a charm :-)

**Comment by RomeuCarvalhoPT** [comment]
Great man, big thanks. I was also doing it, but with my lack of knowledge it was very hard.
I very looking forward to have it, please keep us informed.

**Comment by thallaa** [comment]
Anxiously waiting for the binding to become public, any idea when it would be available?

**Comment by kwave** [comment]
Sorry guys, I've had little time the last few days. The only thing still open to implement is to use the correct representation for all homie properties (ON/OFF should be a switch, not a String). I hope to get this finished no later than the end of next week.

**Comment by MasterIpse** [comment]
@kwave thanks for your work on the binding! This should make the Sonoffs super easy to use!

I built the binding from source, openHAB discovers my Sonoff, but I can not for the life of me get a switch to show up as a channel. 
The device properties like uptime, status etc. work.

Homie on an Sonoff sends following retained mqtt messages for the switch:
/home/device-id/$nodes -> Switch:Switch
/home/device-id/Switch/$type -> ESH:PowerOutlet  (also tried Switch)
/home/device-id/Switch/$properties -> value:settable,value,itemtype
/home/device-id/Switch/value -> OFF
/home/device-id/Switch/itemtype -> Switch

I also tried the heater example with the simulateHomieMQTT.js, the "Wemos" are discovered but there is no "heater" channel.
Am I doing this all wrong? Any help would be greatly appreciated!

**Comment by ThomDietrich** [comment]
This might be a bit off-topic. You should look for help in the openHAB community forum or in the PR (if this might be a bug), e.g. 

* https://github.com/openhab/openhab2-addons/pull/1741
* https://community.openhab.org/t/framework-homie-for-esp8266/8674
* https://community.openhab.org/t/itead-sonoff-switches-and-sockets-cheap-esp8266-wifi-mqtt-hardware/15024

**Comment by kwave** [comment]
@MasterIpse  As this is really kinda off-topic here, I'll contact you via mail.

**Comment by JannikJung0** [comment]
@kwave Is there any news?
I checked your repository, but couldn't find the binding.

I would be really interested in testing it out!

**Comment by jmvaz** [comment]
@kwave Any news with the binding? where we can take one shapshot version to start testing?
Regards

**Comment by kwave** [comment]
The binding is ready and already working. I am still waiting for the OH guys to approve my PR.

**Comment by jmvaz** [comment]
Can you provide one jar for testing?

**Comment by maurostaiti** [comment]
Hi Kwave,
I have the same problem as MasterIpse....can you help me please?

**Comment by timpur** [comment]
Thought i'd share this for people looking for it.
https://github.com/kwave/openhab2-addons/tree/bindings/homie/addons/binding/org.openhab.binding.homie

**Comment by timpur** [comment]
@kwave Any news?
Also do you have a jar for me to get the addon working ? And also how to install it in openhab2 ?

Thanks.

**Comment by sja** [comment]
Have you seen my comment on the OH PR? I have problems using my nodes. Would be great if we have a working example.

**Comment by davidgraeff** [comment]
I was going to implement a binding for easyesp, the only other mqtt convention that I know but was highly unsatisfied with the oh/eclipse smart home mqtt transport bundle.

I refactored it and my changes got merged already. I also add a binding to configure brokers and a binding for a simple mqtt topic to thing mapping. With those merged it should be much easier to develop mqtt bindings. 

This new homie binding probably needs adaptation though.

**Comment by ThomDietrich** [comment]
For reference:
* https://github.com/eclipse/smarthome/issues/3812
* https://github.com/eclipse/smarthome/pull/3876

**Comment by euphi** [comment]
Is there any news? If there would be a JAR, I would give it a try.

**Comment by davidgraeff** [comment]
While the homie binding was developed, the underlying Mqtt subsystem changed. The binding is not yet adapted by the author, so no progress here unfortunately.

**Comment by dhopkins1000** [comment]
Hi @kwave. Have you been able to update your homie binding and re-publish yet ?
I started integrating a few mqtt nodes based on homie and it would be great if I could operate them through openhab2.

**Comment by ThomDietrich** [comment]
@dhopkins1000

The progress of the binding can be seen here: https://github.com/openhab/openhab2-addons/pull/1741

While it is sadly not yet available you can still operate your Homie device through openHAB with the general MQTT binding. See: https://community.openhab.org/t/mqtt-binding-v1-11-getting-started-101/33958

**Comment by jmvaz** [comment]
@kwave you plan to finish this binding? I checked and you have been away from openhab forum since may this year. :(

**Comment by ThomDietrich** [comment]
See #85

**Comment by ThomDietrich** [comment]
Closing in favor of #85 

Please comment if you disagree.
If an argument here is still relevant, consider bringing it up in the new thread.
